### PR TITLE
Migrate groups API to fastAPI

### PR DESCRIPTION
**Pull request by arash77**
## What did you do?

- Include an Integration test under `test/integration/`
- Refactor the legacy `groups` API logic so all the logic is contained in the `GroupsManager` class
- Create pydantic models for the responses and payloads
- Create the FastAPI routes that will replace the old ones

related to: https://github.com/galaxyproject/galaxy/issues/10889

## Why did you make this change?
This is part of the efforts to migrate the full Galaxy API to FastAPI see #10889.

## How to test the changes?
- [x] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

### CODE REVIEW

File: lib/galaxy/managers/groups.py [hunk 1]
```diff
@@ -51,21 +48,21 @@ def index(self, trans: ProvidesAppContext):
...
-        name = payload.get("name", None)
+        name = payload.name
         if name is None:
             raise ObjectAttributeMissingException("Missing required name")
         self._check_duplicated_group_name(sa_session, name)
 
         group = model.Group(name=name)
         sa_session.add(group)
-        encoded_user_ids = payload.get("user_ids", [])
+        encoded_user_ids = getattr(payload, "user_ids", [])
```
Comment (davelopez) [inline lib/galaxy/managers/groups.py]: Why use `getattr` here? payload seems to be now a Pydantic model.

File: lib/galaxy/schema/groups.py [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+
+class GroupCreatePayload(Model):
+    """Payload schema for creating a group."""
+
+    name: str = Field(
+        Required,
+        title="name of the group",
+        description="name of the group",
+    )
+    user_ids: List[EncodedDatabaseIdField] = Field(
```
Comment (davelopez) [inline lib/galaxy/schema/groups.py]: For `input` models that contain database IDs you should use `DecodedDatabaseIdField` this will handle the validation and decoding of the ID automatically and you can directly access the decoded ID (int) when using them.

File: lib/galaxy/webapps/galaxy/api/groups.py [hunk 3]
```diff
@@ -2,61 +2,85 @@
...
+    @router.get(
+        "/api/groups",
+        summary="Displays a collection (list) of groups.",
+        require_admin=True,
+    )
+    def index(
+        self,
+        trans: ProvidesAppContext = DependsOnTrans,
+    ) -> GroupIndexListResponse:
+        """GET /api/groups - Displays a collection (list) of groups."""
```
Comment (davelopez) [inline lib/galaxy/webapps/galaxy/api/groups.py]: These docstrings will appear in the interactive documentation under `http://127.0.0.1:8080/api/docs` once you view the details of that particular route. You should use this to write detailed documentation of the endpoint if needed, but in this case, it is probably fine to just remove it if it's not adding anything to the summary.

The same for the rest of the endpoints.

File: lib/galaxy/webapps/galaxy/api/groups.py [hunk 4]
```diff
@@ -2,61 +2,85 @@
...
+        summary="Displays information about a group.",
+        require_admin=True,
+    )
+    def show(
+        self,
+        trans: ProvidesAppContext = DependsOnTrans,
+        encoded_group_id: EncodedDatabaseIdField = Path(),
+    ) -> GroupShowResponse:
+        """GET /api/groups/{encoded_group_id} - Displays information about a group."""
+        group_id = EncodedDatabaseIdField.decode(encoded_group_id)
```
Comment (davelopez) [inline lib/galaxy/webapps/galaxy/api/groups.py]: Use a `DecodedDatabaseIdField` for any input parameter or payload model field, this will validate and auto-decode the ID so you don't need to manually decode it like here.

File: test/integration/test_groups.py [hunk 5]
```diff
@@ -0,0 +1,122 @@
+from typing import Optional
+
+from galaxy_test.base.populators import DatasetPopulator
+from galaxy_test.driver import integration_util
+
+
+class TestGroupsApi(integration_util.IntegrationTestCase):
```
Comment (davelopez) [inline test/integration/test_groups.py]: There seem to be existing API tests for groups in `lib/galaxy_test/api/test_groups.py` and there is no configuration change in Galaxy needed, so these integration tests are unnecessary.

File: lib/galaxy/schema/groups.py [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+)
+
+from galaxy.schema.fields import EncodedDatabaseIdField
+from galaxy.schema.schema import Model
+
+
+class GroupIndexResponse(Model):
+    """Response schema for a group."""
+
+    model_class: str = Field(
```
Comment (davelopez) [inline lib/galaxy/schema/groups.py]: The `model_class` here is not just a string, is a particular literal and must always be the same. Check out https://github.com/galaxyproject/galaxy/blob/05fe2868f3a8af1b634579f5c3ca2b47d538e357/lib/galaxy/schema/schema.py#L431 you can also find that there are existing models that you can reuse for groups. You can move them to /galaxy/schema/groups.py` if you want.

**Review by davelopez — COMMENTED** [review]
Nice first attempt!

Find a couple of comments below explaining a bit some of our conventions.

File: lib/galaxy/managers/groups.py (hunk 1 above)
Comment (arash77) [reply lib/galaxy/managers/groups.py]: Thanks for the feedback. I've updated the pull request with the modifications.

File: lib/galaxy/webapps/galaxy/api/groups.py (hunk 3 above)
Comment (davelopez) [reply lib/galaxy/webapps/galaxy/api/groups.py]: You can completely remove the docstrings line if there is no detailed docs for he endpoint :)

File: lib/galaxy/managers/groups.py [hunk 7]
```diff
@@ -119,25 +112,17 @@ def _get_group(self, sa_session: galaxy_scoped_session, group_id: int) -> model.
             raise ObjectNotFound("Group with the provided id was not found.")
         return group
 
-    def _get_users_by_encoded_ids(
-        self, sa_session: galaxy_scoped_session, encoded_user_ids: List[EncodedDatabaseIdField]
+    def _get_users_by_ids(
+        self, sa_session: galaxy_scoped_session, user_ids: List[DecodedDatabaseIdField]
     ) -> List[model.User]:
-        user_ids = self._decode_ids(encoded_user_ids)
         return get_users_by_ids(sa_session, user_ids)
```
Comment (davelopez) [inline lib/galaxy/managers/groups.py]: Looks like this function is no longer relevant and can be replaced directly by `get_users_by_ids`

File: lib/galaxy/managers/groups.py [hunk 8]
```diff
@@ -119,25 +112,17 @@ def _get_group(self, sa_session: galaxy_scoped_session, group_id: int) -> model.
...
 
-    def _get_roles_by_encoded_ids(
-        self, sa_session: galaxy_scoped_session, encoded_role_ids: List[EncodedDatabaseIdField]
+    def _get_roles_by_ids(
+        self, sa_session: galaxy_scoped_session, role_ids: List[DecodedDatabaseIdField]
     ) -> List[model.Role]:
-        role_ids = self._decode_ids(encoded_role_ids)
         stmt = select(Role).where(Role.id.in_(role_ids))
         return sa_session.scalars(stmt).all()
```
Comment (davelopez) [inline lib/galaxy/managers/groups.py]: Similar to `get_users_by_ids` this function can be moved to `lib/galaxy/managers/roles.py`

File: lib/galaxy/schema/groups.py [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+
+GROUP_MODEL_CLASS = Literal["Group"]
+
+
+class GroupIndexResponse(Model):
+    """Response schema for a group."""
+
+    model_class: GROUP_MODEL_CLASS = ModelClassField(GROUP_MODEL_CLASS)
+    id: DecodedDatabaseIdField = Field(
+        ...,
```
Comment (davelopez) [inline lib/galaxy/schema/groups.py]: It is fine to use the `Required` constant instead of the ellipsis (...) for required fields, the reason the ellipsis was used in all the previous models is because the constant was not available in older versions of Pydantic.

File: lib/galaxy/schema/groups.py [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+from galaxy.schema.schema import Model
+
+GROUP_MODEL_CLASS = Literal["Group"]
+
+
+class GroupIndexResponse(Model):
+    """Response schema for a group."""
+
+    model_class: GROUP_MODEL_CLASS = ModelClassField(GROUP_MODEL_CLASS)
+    id: DecodedDatabaseIdField = Field(
```
Comment (davelopez) [inline lib/galaxy/schema/groups.py]:
```suggestion
    id: EncodedDatabaseIdField = Field(
```

Sorry, I was not completely clear on how to use the Encoded/Decoded ID handling in models. For **response** models we always use the `Encoded` version of the DatabaseIdField, but for **request** (body, query, or path parameters) we use the `Decoded` version. This ensures that we validate incoming IDs to be properly encoded and provide the decoded value directly for use on the endpoints without manual decoding involved and also ensures that the serialized models in the response are encoded properly when they are returned.

So in summary:
- Input -> Decoded
- Output -> Encoded

File: lib/galaxy/schema/groups.py [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+    )
+
+
+class GroupIndexListResponse(Model):
+    """Response schema for listing groups."""
+
+    __root__: List[GroupIndexResponse]
+
+
+class GroupShowResponse(Model):
```
Comment (davelopez) [inline lib/galaxy/schema/groups.py]: This model name seems very coupled with the "show" endpoint, but it can be used in other endpoints like the "create".

File: lib/galaxy/schema/groups.py [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+    """Response schema for showing a group."""
+
+    model_class: GROUP_MODEL_CLASS = ModelClassField(GROUP_MODEL_CLASS)
+    id: str = Field(
```
Comment (davelopez) [inline lib/galaxy/schema/groups.py]:
```suggestion
    id: EncodedDatabaseIdField = Field(
```

File: lib/galaxy/schema/groups.py (hunk 11 above)
Comment (arash77) [reply lib/galaxy/schema/groups.py]: I'm a bit confused. Isn't `GroupShowResponse` exclusively for the 'show' endpoint because of `roles_url` and `users_url`?

File: lib/galaxy/managers/groups.py (hunk 7 above)
Comment (arash77) [reply lib/galaxy/managers/groups.py]: Thank you for the review. I will apply them ASAP.

File: lib/galaxy/webapps/galaxy/api/groups.py [hunk 13]
```diff
@@ -2,61 +2,79 @@
...
+    @router.post(
+        "/api/groups",
+        summary="Creates a new group.",
+        require_admin=True,
+    )
+    def create(
+        self,
+        trans: ProvidesAppContext = DependsOnTrans,
+        payload: GroupCreatePayload = Body(...),
+    ) -> GroupIndexListResponse:
```
Comment (davelopez) [inline lib/galaxy/webapps/galaxy/api/groups.py]: I was referring to this, the create endpoint returns a single Group response so the same response model is used for the `show` and `create` edpoints.

```suggestion
    ) -> GroupShowResponse:
```

File: lib/galaxy/schema/groups.py (hunk 11 above)
Comment (davelopez) [reply lib/galaxy/schema/groups.py]: Sorry I was not very clear, see this comment https://github.com/galaxyproject/galaxy/pull/17051/files#r1400813994

I was just suggesting that maybe `GroupResponse` is a more generic name because `Show` is very tied to one of the endpoints.

File: lib/galaxy/webapps/galaxy/api/groups.py (hunk 13 above)
Comment (arash77) [reply lib/galaxy/webapps/galaxy/api/groups.py]: I used `GroupIndexListResponse` for `create` endpoint because it returns a list with a single group response. but I can make it to not create a list but just a group response.

Comment (davelopez) [reply lib/galaxy/webapps/galaxy/api/groups.py]: I was not expecting a list here... I wonder if it's a bug I introduced or if it was like that in the original API so we need to keep it for backward compatibility I can't remember it... :sweat_smile:

Comment (arash77) [reply lib/galaxy/webapps/galaxy/api/groups.py]: It was a list from the original API:
https://github.com/galaxyproject/galaxy/blob/e8a288b2d389b756c651cbe1a5682239dd6233fe/lib/galaxy/managers/groups.py#L77

Comment (davelopez) [reply lib/galaxy/webapps/galaxy/api/groups.py]: Yes, I checked back until [the original API endpoint was created in 2012](https://github.com/galaxyproject/galaxy/commit/9bcbabfaa9d1a8a70d4e6e7155015a3a4c8d8c97#diff-f36fa9ebf45ae5ff33bbf8f2749b36b9677ee1fb1600026e61d1c68eabf3059dR70) and it also returns a list with one single group... weird... but I guess it needs to be backward compatible, so all good :+1: 

Sorry for the noise!

File: lib/galaxy/managers/groups.py [hunk 14]
```diff
@@ -47,33 +35,33 @@ def index(self, trans: ProvidesAppContext):
         for group in get_not_deleted_groups(trans.sa_session):
             item = group.to_dict(value_mapper={"id": DecodedDatabaseIdField.encode})
             encoded_id = DecodedDatabaseIdField.encode(group.id)
-            item["url"] = url_for("group", id=encoded_id)
+            item["url"] = self._url_for(trans, "index", group_id=encoded_id)
```
Comment (davelopez) [inline lib/galaxy/managers/groups.py]: These URLs don't look correct. It should be returning something like: `/api/groups/{encoded_group_id}` but this one is returning `/api/groups/{encoded_group_id}/roles`. You need to play with the `name` of the FastAPI route to make it generate the route you need.

File: lib/galaxy/managers/groups.py [hunk 15]
```diff
@@ -83,32 +71,48 @@ def show(self, trans: ProvidesAppContext, group_id: int):
...
+        encoded_user_id = [DecodedDatabaseIdField.encode(gu.user.id) for gu in group.users]
+        encoded_role_id = [DecodedDatabaseIdField.encode(gr.role.id) for gr in group.roles]
+        item["users_url"] = [
+            self._url_for(trans, "group_user", group_id=encoded_id, user_id=group_user)
+            for group_user in encoded_user_id
+        ]
+        item["roles_url"] = [
+            self._url_for(trans, "group_role", group_id=encoded_id, role_id=group_role)
+            for group_role in encoded_role_id
+        ]
```
Comment (davelopez) [inline lib/galaxy/managers/groups.py]: The same for these URLs, they should not be lists but rather something like:
`users_url: "/api/groups/{encoded_group_id}/users"`

I also think some of those URLs are currently wrong in the `dev` branch which kind of confirms that nobody is using them... so we should probably deprecate them and remove them at some point.

**Review by davelopez — COMMENTED** [review]
Almost there! Pretty good so far!

**Review by davelopez — APPROVED** [review]
Awesome work!
Congratulations on your first contribution to Galaxy! :rocket:

**Comment by davelopez** [comment]
Looks like there are some conflicts now, I recommend you use rebase to fix the conflict instead of merging the dev branch to have a cleaner list of commits.

**Comment by davelopez** [comment]
Test failures seem unrelated and fail intermittently in other PRs too.

Thank you @arash77 for addressing all the comments! :+1:
